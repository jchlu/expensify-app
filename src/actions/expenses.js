import uuid from 'uuid'
import database from '../firebase/firebase'
// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  // Returning a function here only works because of thunk middleware
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }
    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const reference = `users/${uid}/expenses/${id}`
    return database.ref(reference).remove().then(() => {
      dispatch(removeExpense({ id }))
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const reference = `users/${uid}/expenses/${id}`
    return database.ref(reference).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    // Fetch all expenses data once from firebase
    return database.ref(`users/${uid}/expenses`)
      .once('value')
      .then((snapshot) => {
      // Parse the data into an array (as per firebase.js testing)
        const expenses = []
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        // Dispatch SET_EXPENSES with firebase array
        dispatch(setExpenses(expenses))
      })
      .catch((error) => {
        console.log('Error fetching data', error)
      })
  }
}
