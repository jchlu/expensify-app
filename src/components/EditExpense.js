import React from 'react'
// Internal Components

export default class EditExpense extends React.Component {
  state = {
    options: []
  }

  handleState = () => {
    this.setState(() => ({ options: [] }))
  }

  render() {
    const title = 'Edit Expense'
    return (
      <div>
        <div className="container">
          <p>{title}</p>
        </div>
      </div>
    )
  }
} // End of EditExpense class definition
