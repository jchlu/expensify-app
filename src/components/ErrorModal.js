import React from 'react'
// Third Party Components
import Modal from 'react-modal'

/** ShowErrorModal requires three params,
 * a message, a show boolean and a function to call on close */
const ErrorModal = (props) => (
  <Modal
    isOpen={!!props.showError}
    onRequestClose={props.handleClearShowError}
    contentLabel="Error Message"
    closeTimeoutMS={300}
    className="modal"
    appElement={document.getElementById('app')}
  >
    <h3 className="modal__title">Warning</h3>
    {props.errorMessage && <p className="modal__body">{props.errorMessage}</p>}
    <button className="button" onClick={props.handleClearShowError}>Okay</button>
  </Modal>
)

export default ErrorModal
