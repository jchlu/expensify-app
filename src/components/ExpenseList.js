import React from 'react'
import { connect } from 'react-redux'

const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    <p>{props.filters.text}</p>
    {props.expenses.length}
  </div>
)

/** What we want from the store to be passed as props */
const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
    filters: state.filters
  }
}

export default connect(mapStateToProps)(ExpenseList)

/*
// *** Original iteration: ***
const ConnectedExpenseList = connect((state) => {
  return {
    expenses: state.expenses
  }
})(ExpenseList)

export default ConnectedExpenseList

// *** Original becomes: ***
export default connect((state) => {
  return {
    expenses: state.expenses
  }
})(ExpenseList)

// *** Which then becomes: ***
const mapStateToProps = (state) => {
  return {
    expenses: state.expenses
  }
}
export default connect(mapStateToProps)(ExpenseList)
 */
