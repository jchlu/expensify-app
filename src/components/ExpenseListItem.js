import React from 'react'

export default ({ description, amount, createdAt }) => (
  <div>
    <p>{description}</p>
    <p>{amount} - {createdAt}</p>
  </div>
)
