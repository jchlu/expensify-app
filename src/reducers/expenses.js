/** expensesReducer */

const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      // return state.concat(action.expense)
      // Rather than concat, use the spread operator
      return [...state, action.expense]
    case 'REMOVE_EXPENSE':
      // Deconstruct the input array as we only need id, implicitly return boolean ~{}~
      return state.filter(({ id }) => id !== action.id)
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          }
        } else {
          return expense
        }
      })
    default:
      return state
  }
}
