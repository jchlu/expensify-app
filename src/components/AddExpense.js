import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

/* Avoid inline functions by switching to a class
* exporting the unconnected version to allow testing */
export class AddExpense extends React.Component {
  onSubmit = (expense) => {
    this.props.onSubmit(expense)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
      <p>Add Expense</p>
      <ExpenseForm
        onSubmit={this.onSubmit}
      />
    </div>
  )
  }
}

/* const AddExpense = (props) => (
  <div>
    <p>Add Expense</p>
    <ExpenseForm
      onSubmit={(expense) => {
        props.onSubmit(expense)
        props.history.push('/')
      }}
    />
  </div>
) */

const mapDispatchToProps = (dispatch) => ({
  // return dispatcher functions, abstracting them away from the component
  onSubmit: (expense) => dispatch(addExpense(expense))
})
export default connect(undefined, mapDispatchToProps)(AddExpense)
