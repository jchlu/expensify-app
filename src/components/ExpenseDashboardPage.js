import React from 'react'
import ExpenseList from './ExpenseList'
// Internal Components

const ExpenseDashboardPage = () => (
  <div>
    <h1><ExpenseList /></h1>
  </div>
)

export default ExpenseDashboardPage
