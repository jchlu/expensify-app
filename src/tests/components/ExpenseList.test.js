import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses'

test('Should render list with fixtures', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />)
  expect(wrapper).toMatchSnapshot()
})

test('Should render empty message', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)
  expect(wrapper).toMatchSnapshot()
})
