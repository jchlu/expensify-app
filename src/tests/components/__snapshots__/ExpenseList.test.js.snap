// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render empty message 1`] = `
<div>
  <p>
    No expenses to display
  </p>
</div>
`;

exports[`Should render list with fixtures 1`] = `
<div>
  <ExpenseListItem
    amount={150}
    createdAt={0}
    description="Gum"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={15550}
    createdAt={-345600000}
    description="Credit Card"
    id={2}
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={345600000}
    description="Rent"
    id={3}
    key="3"
    note=""
  />
</div>
`;
